php/libraries/User.class.inc

```diff
@@ -525,6 +525,55 @@ class User extends UserPermissions
         $this->update($updateArray);
     }
 
+    /**
+     * Use the Pwned Passwords API to determine whether a user is trying
+     * to use a password that has been revealed in public data breaches.
+     *
+     * @param string $password The password to check for pwnage.
+     *
+     * @return bool Whether the password is known to be pwned.
+     */
+    function pwnedPassword(string $password): bool
+    {
+        // TODO Use the HttpClient in LORIS core to cURL more simply.
+
+        /* The pwned passwords API searches a range of password hashes to
+         * protect the actual SHA1 hash from being revealed.
+         * More here: <https://haveibeenpwned.com/API/v2#PwnedPasswords>
+         */
+        $hash       = sha1($password);
+        $hashPrefix = substr(sha1($password), 0, 5);
+        $url        = "https://api.pwnedpasswords.com/range/$hashPrefix";
+
+        /* Build curl */
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, $url);
+
+        // get response
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+
+        $response = curl_exec($ch);
+        curl_close($ch);
+
+        $hashSuffixes = explode("\n", $response);
+        /* The response will be one long string consisting of suffixes matching
+         * our hash prefix, followed by a colon and the number of times the
+         * password has occurred in data breaches.
+         * Here we iterate over all of the suffixes and check whether the prefix
+         * and the whole suffix line together contain $hash as a substring at
+         * the beginning of the line (instead of preprocessing by removing
+         * the colon and count from the end of the $suffix line.
+         * Check the above link if you want more info.
+         */
+        foreach ($hashSuffixes as $suffix) {
```

Comment:
What about a password like 'vDg62fF6$gN3&dsKe!dfw' that might have been pwn only once? If we remove all PWD that had been pwn, we will end up blocking a lot of them and eventually it will be difficult to find one that have not been pwn.
A quick extrapolation with 3 test would eliminate about 189M passwords, this would just increased in size over time.

Maybe combine the number of pwn and strength together

Reply:
I think the example you gave is a bit of an edge case, and someone with that password is likely to be using a password generator to create it. It's easy to click the button again and generate a new string of random characters.

Even eliminating 189M doesn't put a dent in the entropy of possible passwords. The end goal is to allow any password that is greater than 8 characters that consists of any printable ASCII characters and isn't silly. There's plenty of room to make a good password with that criteria.